soh/src/code/z_parameter.c

```diff
@@ -1690,12 +1690,35 @@ u8 Item_Give(GlobalContext* globalCtx, u8 item) {
         }
         return ITEM_NONE;
     } else if (item == ITEM_KEY_SMALL) {
-        if (gSaveContext.inventory.dungeonKeys[gSaveContext.mapIndex] < 0) {
-            gSaveContext.inventory.dungeonKeys[gSaveContext.mapIndex] = 1;
-            return ITEM_NONE;
+        // Small key exceptions for rando with keysanity off.
+        if (gSaveContext.n64ddFlag) {
+            if (globalCtx->sceneNum == 10) { // ganon's tower -> ganon's castle
+                if (gSaveContext.inventory.dungeonKeys[13] < 0) {
+                    gSaveContext.inventory.dungeonKeys[13] = 1;
+                    return ITEM_NONE;
+                } else {
+                    gSaveContext.inventory.dungeonKeys[13]++;
+                    return ITEM_NONE;
+                }
+            }
+
+            if (globalCtx->sceneNum == 92) { // Desert Colossus -> Spirit Temple.
+                if (gSaveContext.inventory.dungeonKeys[6] < 0) {
+                    gSaveContext.inventory.dungeonKeys[6] = 1;
+                    return ITEM_NONE;
+                } else {
+                    gSaveContext.inventory.dungeonKeys[6]++;
+                    return ITEM_NONE;
+                }
+            }
         } else {
-            gSaveContext.inventory.dungeonKeys[gSaveContext.mapIndex]++;
-            return ITEM_NONE;
+            if (gSaveContext.inventory.dungeonKeys[gSaveContext.mapIndex] < 0) {
+                gSaveContext.inventory.dungeonKeys[gSaveContext.mapIndex] = 1;
+                return ITEM_NONE;
+            } else {
+                gSaveContext.inventory.dungeonKeys[gSaveContext.mapIndex]++;
+                return ITEM_NONE;
+            }
```

Comment:
just a note that i've verified this matches what decomp has for vanilla key logic https://github.com/zeldaret/oot/blob/fe8d5988b909e0c26c41933f8ee2044ca7580fc0/src/code/z_parameter.c#L1429-L1436

Reply:
Oh yeah I actually didn't change that, I think the diff didn't register the else that is being used to differentiated between rando and not as a change, so it split that up and shows it as two different removals when it wasn't. Oh but I think I did just identify a problem here, that else is never gonna run in rando like it should when the key isn't in one of those excepted locations. Should just be able to fix it by removing the else now that I think about it. I'll make a follow up PR shortly.